### PR TITLE
fix: stop duplicating identical interval tables across font styles

### PR DESCRIPTION
**Pull request by eszter007**
## Summary

* **What is the goal of this PR?** Cut the peak RAM cost of loading a multi-style `.cpfont` by not allocating a second copy of a codepoint interval table that an earlier style already has.
I had people on my Japanese-focused fork matcha-reader having heap problems with CJK fonts with size 20 and up, e.g. the book crashing.

* **What changes are included?**
  * `SdCardFont::load()` compares each style's on-disk interval table against every earlier-loaded style's resident table — same count, same encoding, byte-identical content — and aliases the earlier table on a match instead of allocating its own.
  * The comparison streams the file in 16-record stack batches, so it never needs a second table's worth of heap to decide.
  * `PerStyle` gains `intervalsShared` so `freeStyleAll()` never `delete[]`s a borrowed pointer; only the style that allocated the table frees it.

## Root cause

`load()` allocated an independent interval table per present style, even when styles cover the same codepoints. Regular/bold/italic weights of the same family typically share a cmap, so a later style's table is usually a byte-for-byte copy of an earlier one's.

**Comparing before allocating is the part that matters.** De-duplicating after the fact still costs a peak of one table per style, and the peak is what the allocator has to satisfy.

**How big the saving is.** Each style that shares a table avoids allocating one: 6 B per interval for BMP fonts. The size therefore depends on the loader's interval limit, which differs between this repository and the fork where the change was first developed:

| | Interval limit (`MAX_INTERVALS`) | Largest possible saving per extra style | Measured |
|---|---:|---:|---|
| **Upstream (this PR)** | 4,096 | 4,096 × 6 = **24,576 B** | **1,728 B** total on X3: four Pretendard sizes, each sharing a 72-interval table (72 × 6 = 432 B each), measured by @serialx<br><br> // **24,426 B** for one sideloaded 2-style NotoSansJP 16 pt font (4,071 intervals × 6 B) in the X4 Pro simulator<br><br> // A/B on X4, develop `4b17a7bb` vs this PR `241bbbf6`, same book and NotoSansJP 16 pt font, cold open + indexing: `Style 1: sharing style 0's 4071-interval table (24426 B not allocated)`. Allocation failures **686 → 0**, lowest free heap **2,668 B → ~15,000 B**, worst largest free block **2,804 B → 18,420 B** |
| **Downstream (matcha-reader fork)** | 16,384 | 16,384 × 6 = 98,304 B | **26,592 B** on X4: a 2-style NotoSansJP 20 pt font with 4,432 intervals (4,432 × 6 B) |

Latin and Korean fonts carry tens of intervals (all Hangul syllables fall in one contiguous block), so upstream they save a few hundred bytes each. A Japanese or Chinese font near the limit saves about 24 KB. The case this PR is for is a font near the 4,096 limit. There the heap may fit one 24,576-byte table but not two: the old loader rejects the font, and this one loads it. @serialx reproduced exactly that with a limited allocation budget: the baseline failed the duplicate load and this PR succeeded.

**Sideloaded Japanese font, upstream build.** I built NotoSansJP 16 pt with this repository's `fontconvert_sdcard.py` (`--intervals cjk,ascii`), with the Regular file for both styles, and loaded it from `/fonts/` in the iOS simulator on this PR's head. Two weights from the same release normally cover the same characters, so their tables match the same way. The log shows:

```
[SDCF] Style 1: sharing style 0's 4071-interval table (24426 B not allocated)
```

The book (森鴎外, ヰタ・セクスアリス) rendered correctly with the shared table: kanji, kana, furigana and embedded Latin, with no errors logged. The simulator has plenty of heap, so this confirms that sharing happens and that rendering is correct. It does not reproduce the constrained-heap case; @serialx's allocation-budget test above covers that.

One related limit: with the full `cjk,latin-ext,punctuation` preset set, the same font comes to 4,106 intervals. That is over `MAX_INTERVALS`, so the loader rejects it before sharing is reached. Staying under the limit meant trimming the Latin ranges.

The downstream NotoSansJP font is over the upstream limit, so this repository rejects it before sharing happens. Its 26,592 B figure is only reachable in the fork.

## Scope Check

CrossPoint is intentionally narrow. See [SCOPE.md](../blob/master/SCOPE.md) and [ROADMAP.md](../blob/master/ROADMAP.md).
Please confirm:

- [x] I have read SCOPE.md and ROADMAP.md.
- [x] This PR is **not** a new built-in theme (themes are temporarily closed pending the move to SD-loaded themes).
- [x] This PR is **not** a new external network connector (sync engine, cloud storage, remote file access, etc.).
- [x] This PR is **not** an interactive app, writing tool, RSS/news/browser, media playback, or PDF feature.
- [x] The stock firmware does not already handle this well — this is a bug fix to existing SD font loading, not a new feature.
- [x] If this PR touches `freeink-sdk/`, `lib/hal/`, the bootloader, OTA, or recovery code, I have coordinated with the relevant maintainer.

Touches `lib/EpdFont/SdCardFont.{h,cpp}` only — no `freeink-sdk/`, `lib/hal/`, bootloader, OTA or recovery code. Memory footprint is listed as a current focus area in SCOPE.md.

## Additional Context

* **Memory:** reduces peak and resident RAM for a multi-style SD font with shared coverage by roughly (styles − 1) × table size. up to 24,576 B per extra style upstream (4,096-interval limit), and the peak drops from two tables to one. See the table above for measured figures.
* **Stack:** the comparison uses a 192-byte stack buffer (`EpdUnicodeInterval cmp[16]`, 16 × 12 B). The target compiler's stack-usage output shows `load()` growing from 192 B to 400 B. It stays on the stack on purpose: this code runs when the heap may be too fragmented for a duplicate table, so it must not add a heap allocation. `CMP_BATCH = 4` would cut the buffer to 48 B at the cost of more `read()` calls (the reads come from SdFat's sector cache, so there is no extra SD I/O). Say the word if you'd like that.
* **Flash:** +600 B (measured on X3).
* **No format change.** This only affects what is kept resident after load, not the on-disk `.cpfont` layout, so no cache or file version bump.
* **Aliasing safety** is the area worth reviewing: `intervalsShared` styles are skipped by `freeStyleAll()`, so the table is freed exactly once by its allocating style. Styles are only ever aliased to a lower-indexed style, and `freeAll()` walks all styles, so there is no ordering by which a borrowed pointer outlives its owner.
* The comparison re-seeks to `intervalsFileOffset` when a candidate does not match, so the allocate-and-read path below sees the file positioned exactly as before.

**Origin:** this change and its predecessor were developed and run in the matcha-reader fork; the earlier revision de-duplicated after allocating, and this is the corrected compare-first version.

---

### AI Usage

While CrossPoint doesn't have restrictions on AI tools in contributing, please be transparent about their usage as it
helps set the right context for reviewers.

Did you use AI tools to help write this code? _**YES**_

### CODE REVIEW

File: lib/EpdFont/SdCardFont.cpp [hunk 1]
```diff
@@ -671,7 +676,68 @@ bool SdCardFont::load(const char* path) {
...
+    for (uint8_t k = 0; k < i && !s.intervalsShared; k++) {
+      auto& owner = styles_[k];
+      if (!owner.present || owner.header.intervalCount != s.header.intervalCount) continue;
+      if (owner.intervalsAreBmp16 != canUseBmp16) continue;
+      if (!owner.bmpIntervals && !owner.fullIntervals) continue;
+
+      // Streamed in small batches: the whole point is to not need a second table's worth of
+      // heap, so the comparison buffer stays on the stack and well inside the task's budget.
+      static constexpr uint32_t CMP_BATCH = 16;
+      EpdUnicodeInterval cmp[CMP_BATCH];
```
[the remark this answers is not in the document]
Comment (eszter007) [reply lib/EpdFont/SdCardFont.cpp:696]: Not applying this one — moving the buffer to the heap works against the purpose of this code path.

This comparison exists for the moment the heap **cannot** fit a second interval table (on the measured CJK font, 26,592 B wanted against a ~32 KB largest free block). Adding a heap allocation here introduces a new failure mode at exactly that point. The suggested handling — log and return — would make the whole font fail to load because a 192-byte buffer could not be allocated, which is a regression against today's behaviour, where the font loads if it fits.

The stack buffer is deliberate: it cannot fail, and it is released as soon as the comparison ends.

If reviewers would prefer a smaller frame, the batch size is the right lever rather than the heap: `CMP_BATCH = 4` brings the buffer to 48 B. Sequential reads come out of SdFat's sector cache, so a smaller batch means more `read()` calls but no additional SD I/O. Happy to make that change if wanted.

**Comment by serialx** [comment]
Below is an automated QA report. Wondering if this is expected result, or is there environment differences that make heap savings different (less)?

---

Tested on **Xteink X3** against develop at `4b17a7bb`, with PR head `241bbbf6`. No rendering regression found in the tested workload; interval sharing works and page timings are essentially unchanged.

Both builds used identical serial-control instrumentation and PlatformIO `default` settings. Ran three interleaved cold-cache baseline/candidate pairs using Demian with Pretendard 16 pt in portrait: six warmup turns followed by six measured turns per boot, giving **18 measured renders per build**.

| Measurement | Baseline | PR | Change |
|---|---:|---:|---:|
| Median page drawing | 100 ms | 99 ms | −1 ms |
| Median rendering including refresh | 1,624 ms | 1,618.5 ms | −5.5 ms |
| Median cold layout preparation | 4,718 ms | 4,721 ms | +3 ms |
| Median free heap after measured pages* | 45,676 B | 46,144 B | +468 B |
| Median largest free block* | 21,492 B | 23,540 B | +2,048 B |
| Lowest since-boot free-heap watermark* | 15,288 B | 15,748 B | +460 B |
| Static RAM | 56,696 B | 56,696 B | unchanged |
| Flash | 5,568,435 B | 5,569,035 B | +600 B |

**\* Heap measurements include retained caches and background layout state.** Cache residency differed between runs, and largest-block ranges overlapped, so these results do not establish a general fragmentation improvement.

Candidate logs confirmed **1,728 B of duplicate interval payload avoided**: four loaded Pretendard sizes each shared a 72 × 6-byte table. This is the direct allocation saving; subsequent cache allocations affect the net free-heap result. Target `sizeof(PerStyle)` and `sizeof(SdCardFont)` were unchanged.

Additional checks:

- **6/6 paired book-page captures were byte-identical.**
- Switching between Pretendard, built-in Noto Serif and MaruBuri, then restoring the original font/page, passed on both builds. All **6/6 paired smoke captures were byte-identical**.
- ASan/UBSan checks against the actual loader covered matching/nonmatching tables, BMP/full Unicode, repeated reloads, partial-load cleanup and allocation failure. No sanitizer findings. With an allocation budget allowing only one 24,576-byte interval table, baseline failed the duplicate load and the PR succeeded.
- No candidate errors, panics or watchdog resets occurred in accepted measurement windows. Baseline logged one advance-cache allocation failure and continued with matching output. Both builds logged cover-thumbnail allocation errors outside those windows.

Two review notes remain:

1. The [comparison buffer in `SdCardFont.cpp:695–696`](https://github.com/crosspoint-reader/crosspoint-reader/blob/241bbbf6cd97405891e0b65ebd60e59c873dd86b/lib/EpdFont/SdCardFont.cpp#L695-L696) is **192 B**, rather than the description’s 96 B: 16 entries × 12 bytes. Target compiler stack-usage output shows `load()` increasing from **192 B to 400 B**. No overflow was observed, but the description should reflect this cost.
2. The cited 4,432-interval NotoSansJP example exceeds the [existing 4,096-interval loader limit](https://github.com/crosspoint-reader/crosspoint-reader/blob/241bbbf6cd97405891e0b65ebd60e59c873dd86b/lib/EpdFont/SdCardFont.cpp#L601-L610); both versions reject it before sharing. Please clarify the example’s provenance. The claimed 26.6 KB saving was **not reproduced on hardware**.

Coverage is limited to X3 portrait and the tested fonts; other hardware, broad CJK fonts on-device and long-duration stability remain unverified.

**Comment by eszter007** [comment]
> 2\. The cited 4,432-interval NotoSansJP example exceeds the [existing 4,096-interval loader limit](https://github.com/crosspoint-reader/crosspoint-reader/blob/241bbbf6cd97405891e0b65ebd60e59c873dd86b/lib/EpdFont/SdCardFont.cpp#L601-L610); both versions reject it before sharing. Please clarify the example’s provenance. The claimed 26.6 KB saving was **not reproduced on hardware**.

Yeah, the initial measurement was taken downstream where a larger interval table is supported on a physical X4. I found a way to generate a bigger Noto Sans JP font that fits for testing by dropping latin extended. Proper Japanese and Chinese in crosspoint require a support of larger interval tables (which is a separate issue). Korean doesn't have the problem because the interval table records runs of consecutive codepoints instead of individual characters, so they are smaller.

So the the largest saving I was able to achieve is **24,426 B** in the X4 Pro simulator for edge-case testing. On a normal Japanese test book, **18,420 B** on a X4.

Without this change, a book crashed immediately downstream with the CJK 20 font, upstream larger CJK fonts with big tables like JP/Chinese are just slow and the heap becomes very tight.

I attached the font you can use for testing:
[NotoSansJPTest_16.zip](https://github.com/user-attachments/files/32418387/NotoSansJPTest_16.zip)

**Comment by serialx** [comment]
LGTM.

**What I checked.** This PR makes `SdCardFont::load()` share one codepoint interval table between font styles whose tables are byte-identical, instead of allocating a copy per style. The new risk is ownership: two styles now point at one heap block, and only one may free it. I read the full `load()` path, every free path, and every consumer of the table pointers. The ownership story holds: `freeStyleAll()` is only reached from `freeAll()`, which sweeps all four styles in one pass, so a borrowed pointer never outlives its owner. The tables are read-only after load, so sharing cannot cause the two styles to disagree. The count and encoding gate before the comparison, the re-seek after a mismatch, and the 192 B stack buffer all check out. The A/B numbers on X4 match what I measured on X3.

**One non-blocking observation.** The PR reads each later style's interval table from the SD card more often than it needs to.

For context: for every style, `load()` already had a *validation loop* that streams the style's whole interval table from the card, one row at a time, checking that ranges are sorted and offsets line up. It then seeks back to the start and reads the table a second time into the freshly allocated buffer. That was two passes.

This PR adds a *compare loop* between those two steps. For each earlier style with the same row count, it streams the table again, 16 rows at a time, and checks for equality. If nothing matches, the original allocate-and-read pass still runs afterwards. So a style with a unique table is now read three times, and four times if there are two earlier candidates. For a 3-style CJK font with distinct tables that is about 50 KB of extra reads on the font-switch path.

The comparison can ride inside the validation loop instead, since that loop already holds each row as it goes by. Sketch:

```cpp
// Before the validation loop: every earlier style with the same row count is a candidate.
uint8_t candidates = 0;
for (uint8_t k = 0; k < i; k++)
  if (styles_[k].present && styles_[k].header.intervalCount == s.header.intervalCount)
    candidates |= (1u << k);

// Inside the validation loop, after the existing checks on the row `iv`:
for (uint8_t k = 0; k < i; k++) {
  if (!(candidates & (1u << k))) continue;
  const auto& o = styles_[k];
  const bool same = o.intervalsAreBmp16
      ? (o.bmpIntervals[j].first == iv.first && o.bmpIntervals[j].last == iv.last && o.bmpIntervals[j].offset == iv.offset)
      : (o.fullIntervals[j].first == iv.first && o.fullIntervals[j].last == iv.last && o.fullIntervals[j].offset == iv.offset);
  if (!same) candidates &= ~(1u << k);
}

// After the validation loop: alias the lowest surviving candidate; otherwise seek back and allocate as today.
```

What this changes:

- Reads per style drop from `2 + candidates` passes to at most 2, and to 1 when a match is found.
- The separate 16-row compare buffer goes away, so `load()` returns to roughly its old stack frame. That also settles the `CMP_BATCH = 4` question in the description.
- Peak heap is unchanged. The decision is still made before any allocation, which is the property this PR exists for.
- The `intervalsShared` flag and the `freeStyleAll()` guard stay exactly as they are.
- The compare loop with its own read, batching, and re-seek logic is replaced by a bitmask and one inner loop.

The encoding gate (`intervalsAreBmp16 == canUseBmp16`) becomes redundant rather than wrong: if every row matches, both styles reach the same encoding decision. I would keep it as a defensive check.

Line references on this branch: validation loop `SdCardFont.cpp:641-667`, compare loop `:687-724`, allocate-and-read `:733-770`.

Fine to leave this for a follow-up if you would rather keep the PR as is.

**Review by serialx — APPROVED** [review]

**Comment by serialx** [comment]
@itsthisjustin I think you might want to review this. Looks good to me

**Review by itsthisjustin — APPROVED** [review]

**Comment by Uri-Tauber** [comment]
`SdCardFont::load()` already reads every interval to validate it and choose its resident representation. This PR rewinds the file and adds a second mini-loader that reads the same table once per candidate style. The sharing decision can be accumulated during the existing validation read.

**Comment by eszter007** [comment]
@Uri-Tauber Good point, thanks. Done in d35215a2.

**Review by Uri-Tauber — APPROVED** [review]
